lucene/core/src/java/org/apache/lucene/search/FilteringFieldComparator.java

```diff
@@ -0,0 +1,83 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.lucene.search;
+
+import java.io.IOException;
+
+/**
+ * Decorates a wrapped FieldComparator to add a functionality to skip over non-competitive docs.
+ * FilteringFieldComparator provides two additional functions for a FieldComparator:
+ * 1) {@code competitiveIterator()} that returns an iterator over
+ *      competitive docs that are stronger than already collected docs.
+ * 2) {@code setCanUpdateIterator()} that notifies the comparator when it is ok to start updating its internal iterator.
+ *  This method is called from a collector to inform the comparator to start updating its iterator.
+ */
+public abstract class FilteringFieldComparator<T> extends FieldComparator<T> {
+    final FieldComparator<T> in;
+
+    public FilteringFieldComparator(FieldComparator<T> in) {
+        this.in = in;
+    }
+
+    protected abstract DocIdSetIterator competitiveIterator();
```

Comment:
Let's only have this method on LeafFieldCompatarors, e.g. by doing something like this? FieldComparators are top-level objects so it doesn't make sense to have leaf-level objects defined on them like DocIdSetIterators.

```suggestion
  @Override
  public abstract FilteringLeafFieldComparator getLeafComparator(LeafReaderContext context) throws IOException; // covariant return type

  public static interface FilteringLeafFieldComparator extends LeafFieldComparator {
    DocIdSetIterator competitiveIterator();
  }
```

Then there's a question of whether setCanUpdateIterator should go there too as well or not.

Reply:
Thanks @jpountz, indeed the suggested structure is much better. addressed in 7120424ff